torch/csrc/jit/codegen/cuda/lower_index.cpp

```diff
@@ -102,6 +102,19 @@ void IndexLowering::handle(const TernaryOp* top) {
       top->getTernaryOpType(), out, in1, in2, in3));
 }
 
+void IndexLowering::handle(const ViewAsScalar* uop) {
+  const auto in = lowerSrcIndex(uop->in(), uop->out());
+  const auto out = lowerDstIndex(uop->out());
+  for (auto loop : for_loops_) {
+    if (loop->iter_domain()->sameAs(uop->vector_id())) {
```

Comment:
This is a tricky part. `sameAs` with `IterDomain` doesn't guarantee that loop is the one `vector_id()` corresponds to. 

For example, this would fail:

```
TEST_F(NVFuserTest, FusionViewAsRealComputeAt_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  // TODO: We should modify our schedulers to correctly handle
  // view_as_real. And test these schedulers.
  std::vector<int64_t> input_shape{512};

  TensorView* x =
      makeSymbolicTensor(input_shape.size(), DataType::ComplexFloat);
  TensorView* bias =
      makeSymbolicTensor(input_shape.size(), DataType::ComplexFloat);
  fusion.addInput(x);
  fusion.addInput(bias);

  TensorView* x_add_bias_2 = makeSymbolicTensor(input_shape.size() + 1);
  fusion.addInput(x_add_bias_2);

  auto x_add_bias_2_plus_1 = add(x_add_bias_2, IrBuilder::create<Double>(1));

  auto x_add_bias = add(x, bias);
  auto x_view = view_as_real(x_add_bias);
  auto out = add(x_add_bias_2_plus_1, x_view);
  fusion.addOutput(out);

  out->axis(0)->parallelize(ParallelType::TIDx);
  x_add_bias->computeAt(out, -1);
  x_add_bias_2->computeAt(out, -1);

  fusion.printMath();
  fusion.printKernel();

#if 0
  auto options =
      at::TensorOptions().dtype(at::kComplexFloat).device(at::kCUDA, 0);
  at::Tensor at_x = at::randn(input_shape, options);
  at::Tensor at_bias = at::randn(input_shape, options);
  std::vector<IValue> aten_inputs = {at_x, at_bias};

  FusionExecutor fe;
  fe.compileFusion(&fusion, aten_inputs);
  auto outputs = fe.runFusion(aten_inputs);

  auto at_x_add_bias = at_x + at_bias;
  auto at_x_view = at::view_as_real(at_x_add_bias) * 2;

  testValidate(&fusion, outputs, aten_inputs, {at_x_view}, __LINE__, __FILE__);
#endif
}
```

The correct way is to use the computeAt Parallel map. This should fix the problem:

```
diff --git a/torch/csrc/jit/codegen/cuda/lower_index.cpp b/torch/csrc/jit/codegen/cuda/lower_index.cpp
index 343d8b911a..75069ed65e 100644
--- a/torch/csrc/jit/codegen/cuda/lower_index.cpp
+++ b/torch/csrc/jit/codegen/cuda/lower_index.cpp
@@ -106,11 +106,20 @@ void IndexLowering::handle(const SplitVector* uop) {
   const auto in = lowerSrcIndex(uop->in(), uop->out());
   const auto out = lowerDstIndex(uop->out());
   for (auto loop : for_loops_) {
+#if 0
     if (loop->iter_domain()->sameAs(uop->vector_id())) {
       Val* vector_id = loop->index();
       pushBack(IrBuilder::create<SplitVector>(out, in, vector_id));
       return;
     }
+#else
+    if (GpuLower::current()->caParallelMap().areMapped(loop->iter_domain(),
+                                                       uop->vector_id()->as<IterDomain>())) {
+      Val* vector_id = loop->index();
+      pushBack(IrBuilder::create<SplitVector>(out, in, vector_id));
+      return;
+  }
+#endif
   }
```

Reply:
I have made the change (lowering + test) according to your review. Thank you! I think I roughly understand the idea. Is it because when two computations are sharing the same loop, then the loop's `iter_domain()` will not be the same as the IterDomain of any of these two computations?

Reply:
This doesn't matter when only one tensor is defined in a loop, but computeAt effectively makes multiple tensors share the same loop. In that case, the IterDomain of a loop is the same as that of *one* of the tensors whose defining expressions are located in the loop. In the above fusion example, `x_add_bias_2` and its consumers share the same loop with the VectorComponent IterDomain of `x_view`. There's several rules to select which IterDomain is used to create a loop, but that IterDomain is called a concrete domain. All the IterDomains sharing the same loop can be found in `caParallelMap()`. For each IterDomain, `caParallelMap().getConcreteID(IterDomain*)` returns its concrete IterDomain, which is the one selected as the IterDomain of `ForLoop`. 

So, in this case, we have `ForLoop`, so we know which IterDomain is the concrete domain (i.e., loop->iter_domain() is the concrete domain). But there's no guarantee the concrete domain is picked up from this particular tensor. The only guarantee we have is `caParallelMap().areMapped()` returns true for those domains.